service/history/replication/task_ack_manager.go

```diff
@@ -92,29 +98,44 @@ func (t *TaskAckManager) GetTasks(ctx context.Context, pollingCluster string, la
 		return nil, err
 	}
 
-	var replicationTasks []*types.ReplicationTask
-	readLevel := lastReadTaskID
-TaskInfoLoop:
+	// Happy path assumption - we will push all tasks to replication tasks.
+	replicationTasks := make([]*types.ReplicationTask, 0, len(tasks))
+
+	var (
+		readLevel                      = lastReadTaskID
+		oldestUnprocessedTaskTimestamp = int64(0)
+		oldestUnprocessedTaskID        = int64(0)
+	)
+
 	for _, task := range tasks {
-		replicationTask, err := t.store.Get(ctx, pollingCluster, *task)
+		// it does not matter if we can process task or not, but we need to know what was the latest taskID we have read.
+		oldestUnprocessedTaskID = task.TaskID
+		oldestUnprocessedTaskTimestamp = task.CreationTime
 
-		switch err.(type) {
-		case nil:
-			// No action
-		case *types.BadRequestError, *types.InternalDataInconsistencyError, *types.EntityNotExistsError:
-			t.logger.Warn("Failed to get replication task.", tag.Error(err))
-		default:
-			t.logger.Error("Failed to get replication task. Return what we have so far.", tag.Error(err))
-			hasMore = true
-			break TaskInfoLoop
+		replicationTask, err := t.store.Get(ctx, pollingCluster, *task)
+		if err != nil {
+			if errors.As(err, new(*types.BadRequestError)) ||
+				errors.As(err, new(*types.InternalDataInconsistencyError)) ||
+				errors.As(err, new(*types.EntityNotExistsError)) {
+				t.logger.Warn("Failed to get replication task.", tag.Error(err))
+			} else {
+				t.logger.Error("Failed to get replication task. Return what we have so far.", tag.Error(err))
+				hasMore = true
+				break
+			}
 		}
+
+		// We update readLevel only if we have found matching replication tasks on the passive side.
 		readLevel = task.TaskID
 		if replicationTask != nil {
 			replicationTasks = append(replicationTasks, replicationTask)
 		}
 	}
 	taskGeneratedTimer.Stop()
 
+	t.scope.RecordTimer(metrics.ReplicationTasksLagRaw, time.Duration(t.ackLevels.GetTransferMaxReadLevel()-oldestUnprocessedTaskID))
```

Comment:
I might be missing how this will help with "If passive cluster cannot fetch tasks, does not report replication lag". `oldestUnprocessedTaskID` is almost identical to `readLevel` except when the loop breaks. So it can be off by 1? 
I'd assume we would want to report the lag when `t.reader.Read()` returns err based on `lastReadTaskID`. That part is untouched.

Reply:
Good point. However, the issue was:
We never update readLevel if we fail in t.store.Get (f.e. due to hitting MaxQPS or Cass failures). So, we never reported any lag since the value was never updated. 
I'll try it out and add a a check to the test case to ensure what we get from the metric.

Reply:
I've updated the logic, so we always pick the oldest task from the batch and report no delay if there are no tasks to replicate.

Reply:
should we also report the metrics before returning at line 98?

Reply:
If we fail to read history from the active side, we don't know the delay (as far as I understand), and we should be notified when we fail to pull the history.

Reply:
I see. I thought we could still report lag based on `lastReadTaskID`